File: matlab_src/bisection.m
function result = bisection(fun, left, right, tolerance)
    k = 1;

    while right - left > tolerance
        mid = (right + left) / 2;
        if fun(mid) == 0
            result = mid;
        elseif fun(left) * fun(mid) < 0
            right = mid;
        else
            left = mid;
        end

        result = mid;
        fprintf('%d result: %.14f\n', k, result);
        k = k + 1;
    end
end

File: python_src/bisection.py
def bisection(fun, left ,right, tolerance):
    k = 1
    
    while right - left > tolerance:
        mid = (left + right ) / 2
        if fun(mid) == 0:
            result = mid
        elif fun(left) * fun(mid) < 0:
            right = mid
        else:
            left = mid
        result = mid
        print(f"{k} result: {result}")
        k += 1
        
def equation(x):
    return x - 2**(-x)

bisection(equation, 0, 1, 1e-4)